File: blockchain/src/main/java/pt/um/lei/masb/blockchain/data/LuminosityData.java
package pt.um.lei.masb.blockchain.data;

import pt.um.lei.masb.blockchain.Sizeable;

import javax.persistence.Basic;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import java.math.BigDecimal;
import java.util.Objects;

/**
 * Luminosity data might be output by an ambient light sensor, using lux units
 * or a lighting unit, outputting a specific amount of lumens.
 */
@Entity
public final class LuminosityData extends GeoData implements Sizeable {
    @Id
    @GeneratedValue
    private long id;

    @Basic(optional = false)
    private final BigDecimal lum;

    @Basic(optional = false)
    private final LUnit unit;

    protected LuminosityData() {
        super(new BigDecimal(0), new BigDecimal(0));
        lum = null;
        unit = null;
    }

    public LuminosityData(BigDecimal lum,
                          LUnit unit,
                          BigDecimal lat,
                          BigDecimal lng) {
        super(lat, lng);
        this.lum = lum;
        this.unit = unit;
    }


    /**
     * @return Luminosity reading, either from lighting units or light sensors.
     */
    public BigDecimal getLum() {
        return lum;
    }

    /**
     * @return The unit of measurement, either lumen for lighting units or lux for light sensors.
     */
    public LUnit getUnit() {
        return unit;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        LuminosityData that = (LuminosityData) o;
        return id == that.id &&
                Objects.equals(lum, that.lum) &&
                unit == that.unit;
    }

    @Override
    public int hashCode() {

        return Objects.hash(lum, unit);
    }

    @Override
    public String toString() {
        return "LuminosityData{" +
                "lum=" + lum +
                ", unit=" + unit +
                '}';
    }
}

File: blockchain/src/main/kotlin/pt/um/lei/masb/blockchain/data/LuminosityData.kt
package pt.um.lei.masb.blockchain.data

import com.orientechnologies.orient.core.record.OElement
import pt.um.lei.masb.blockchain.Coinbase
import pt.um.lei.masb.blockchain.Hash
import pt.um.lei.masb.blockchain.persistance.NewInstanceSession
import pt.um.lei.masb.blockchain.utils.Crypter
import java.io.InvalidClassException
import java.math.BigDecimal

/**
 * Luminosity data might be output by an ambient light
 * sensor, using lux units ([LUnit.LUX]) or a lighting unit,
 * outputting a specific amount of lumens ([LUnit.LUMENS]),
 * according to [unit].
 */
data class LuminosityData(
    val lum: BigDecimal,
    val unit: LUnit
) : BlockChainData {
    override fun digest(c: Crypter): Hash =
        c.applyHash(
            """
            $lum
            ${unit.name}
            ${unit.ordinal}
            """.trimIndent()
        )


    override fun store(
        session: NewInstanceSession
    ): OElement =
        session
            .newInstance("Luminosity")
            .let {
                it.setProperty("lum", lum)
                it.setProperty(
                    "unit", when (unit) {
                        LUnit.LUMENS -> 0x00.toByte()
                        LUnit.LUX -> 0x01.toByte()
                    }
                )
                it
            }


    override fun calculateDiff(
        previous: SelfInterval
    ): BigDecimal =
        when (previous) {
            is LuminosityData -> calculateDiffLum(previous)
            else ->
                throw InvalidClassException(
                    "SelfInterval supplied is not ${
                    this::class.qualifiedName
                    }"
                )
        }

    private fun calculateDiffLum(
        previous: LuminosityData
    ): BigDecimal =
        lum.subtract(previous.lum)
            .divide(
                previous.lum,
                Coinbase.MATH_CONTEXT
            )


    override fun toString(): String {
        return "LuminosityData(lum = $lum, unit = $unit)"
    }
}
